### PR TITLE
ente-web: init at 0.9.5

### PR DESCRIPTION
**Pull request by pinpox**
## Description of changes

Part of: https://github.com/NixOS/nixpkgs/issues/292641

Added the web frontend for [ente](https://ente.io). I have packaged the backend in a [previous PR](https://github.com/NixOS/nixpkgs/pull/325659)

~The package works as expected, but is dependant on the new yarn hooks in https://github.com/NixOS/nixpkgs/pull/318015 and therefore marked as draft PR.~

~When https://github.com/NixOS/nixpkgs/pull/318015 gets merged, I will rebase my PR against nixpkgs master and remove the draft bit.~

EDIT: rebased against master.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:26]:
```suggestion
  src = fetchFromGitHub {
    # ...
  } + web;
```

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:13]: This is a bit weird. I'd use `0.9.5`, and adjust the `rev` argument accordingly.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:11]: Remove empty lines.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:49]: Upstream explicitly states this?

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:43]: Nowadays I see many people agreeing that using `with lib` for this scope is too wide, and you can use `lib.licenses` and `with lib.maintainers;`.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:48]: I habven't built this locally, but I will notify you later how to adjust the `installPhase` to improve this situation. I also guess that the resulting `$out` has a weird directory layout structure. You may also want to consider using `npmHooks.npmInstallHook`.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:9]: Could you please use `nixfmt-rfc-style` package?

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:13]: I used the same version/rev for the API client, should I change both then? https://github.com/NixOS/nixpkgs/pull/325659

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:39]: I had a quick look at `npmHooks.npmInstallHook` but didn't get it to work, let me know if you do. The directory structure as it currently is allows hosting the package using something like nginx (see https://wiki.nixos.org/wiki/Ente_Photos for an example).

Let me know if you find a more elegant way to do this

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:40]: Are you refering to the license here? It is agpl3 as per the github page: https://github.com/ente-io/ente

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: done.

**Comment by doronbehar** [comment]
BTW could you please always squash the `ente-web` related commits to a single `init at` commit?

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix:14]: > I used the same version/rev for the API client, should I change both then? #325659

Hmmm... I don't agree with what happened in #325659 . Do you know how to explain upstream's policy regarding their Git tags?

Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:14]: They have multiple apps in the same monorepo, the tags are for the specific software+version if I'm not mistaken. I'm not affiliated with the developers in any way though, so I can just infer that from past public releases.

Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix:14]: > They have multiple apps in the same monorepo, the tags are for the specific software+version if I'm not mistaken. I'm not affiliated with the developers in any way though, so I can just infer that from past public releases.

That's what I thought as well, but what confuses me is that you used the same tag for both the `museum` package and for this package...

Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix:14]: In general, I'd encourage a consistent `src` approach for all packages by the `ente-io` stack.. See also this line from [`pkgs/README.md`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md):

> - The version attribute must start with a digit e.g., "0.3.1rc2".

If possible, I'd use the same `src` for both building `museum` and `ente-web`.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix:50]: > I had a quick look at `npmHooks.npmInstallHook` but didn't get it to work, let me know if you do. The directory structure as it currently is allows hosting the package using something like nginx (see [wiki.nixos.org/wiki/Ente_Photos](https://wiki.nixos.org/wiki/Ente_Photos) for an example).

So this package doesn't generate an executable? Or does nginx uses this executable? And other html/js files? I think I read somewhere you mentioned you have a module sort of working for this, perhaps it'd be better to add it here.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:14]: > That's what I thought as well, but what confuses me is that you used the same tag for both the museum package and for this package...

That is correct I think. The `photos` tags is for both backend (museum) and fronend (ente-web) as far as I can tell. The other tags are `auth` (different software) and `cli` for the command-line tool.

I will change the version attribute to just the digits for ente-web and submit a separate PR for museum

Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: Fixed now. If I should use a different tag, let me know but I think this is correct.

Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix]: > I will change the version attribute to just the digits for ente-web and submit a separate PR for museum

You can add a commit for `museum` here too - it will be easier to review it that way.

**Comment by doronbehar** [comment]
Also don't forget to change the commit message to not include the `photos-v` prefix.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: Will do after rebasing when the hooks are merged. This branch doesn't include museum yet.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:50]: No executable as far as I can tell, just outputs a directory with stuff to put on a webserver. I took inspiration from other packages like zabbix or vaultwarden that have a separate package for the web frontend. I would like to expose it as separate top-level package not tied to the module, since it can make sense to put the front- and backend on different machines.

Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix:50]: > No executable as far as I can tell, just outputs a directory with stuff to put on a webserver.

Then why did you specify a `meta.mainProgram`?

> I took inspiration from other packages like zabbix or vaultwarden that have a separate package for the web frontend. I would like to expose it as separate top-level package not tied to the module, since it can make sense to put the front- and backend on different machines.

I agree with this approach :+1:.

**Comment by supermomme** [comment]
Hey folks,
in the [self-hosting guide](https://help.ente.io/self-hosting/) from ente.io. The api-endpoint is set as an environment variable during `yarn dev` and thus during build time. I haven't seen the environment-variable in your PR. Do you set the api-endpoint after build-time?

**Comment by pinpox** [comment]
> Hey folks, in the [self-hosting guide](https://help.ente.io/self-hosting/) from ente.io. The api-endpoint is set as an environment variable during `yarn dev` and thus during build time. I haven't seen the environment-variable in your PR. Do you set the api-endpoint after build-time?

You can set the endpoint you want to use in the frontend manually by clicking 7 times on the login image (unlocks the setting). But you are right, it would be nice to be able to set it at build time, I just don't know what the best way would be of making that configurable. Maybe let it default to `localhost:8080` and add an override if the user want's to set it?

**Comment by supermomme** [comment]
> You can set the endpoint you want to use in the frontend manually by clicking 7 times on the login image (unlocks the setting). But you are right, it would be nice to be able to set it at build time, I just don't know what the best way would be of making that configurable. Maybe let it default to `localhost:8080` and add an override if the user want's to set it?

I think localhost wouldn't work for public access.
I am not quite good in nix, but maybe some form of [override](https://nixos.org/manual/nixpkgs/stable/#chap-overrides) would work here?

File: pkgs/by-name/en/ente-web/package.nix
Comment (mweinelt) [inline pkgs/by-name/en/ente-web/package.nix:68]:
```
nix-repl> ente-web.meta
{
  available = true;
  broken = false;
  insecure = false;
  name = "ente-web-0.9.5";
  outputsToInstall = [ ... ];
  position = "/home/hexa/git/nixos/master/pkgs/by-name/en/ente-web/package.nix:14";
  unfree = false;
  unsupported = false;
}
```

Not idea what the idea here is, but we're loosing proper metadata on the attribute.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:44]: I don't think this is a good interface to configure this. I think a dedicated NixOS module would be better.

Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:44]: Will revert this then to using a normal, optional override.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: I changed it back to just using an override. If the user chooses to set an override, it might take around ~2 minutes to build without cache. I reconsidered the alternative and think the simpler, more maintainable package is worth the small burden on a user specifically setting this.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: Changed to override, see comment above.

**Comment by pinpox** [comment]
Tested setting the endpoint with:
```
nix build --expr "with import ./. {}; ente-web.overrideAttrs { publicEndpoint = ''test.com''; }" --impure
```

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:41]: ?

How about using only a single `publicEndpoint` argument to the package? It'd also be nice to document the default publicEndpoint using the `?` default argument syntax.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:12]: The whole point of being able to pass a function to `stdenv.mkDerivation`, is to not use `rec`:

```suggestion
stdenv.mkDerivation (finalAttrs: {
```

Use `finalAttrs.version` (for example) when needed - like you use `finalAttrs.src`.

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:10]:
```suggestion
```

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:11]:
```suggestion
}:

```

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:40]: I'm still working this out, that's why I marked the PR as WIP. It seems the build accepts more env vars and the build procces does different things depending on which of those are set. The defaults behaviour is when nothing is set, so I think it makes sense to make the whole env default to not being set.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: Museum fixed in separate commit.

**Comment by pinpox** [comment]
I think this is ready for review. I addressed all concerns and squashed into two commits. I'm running this package myself for a while now and it works fine. The `env` allow specifying arbitrary environment vars, which I think is the best way to go here. Multiple settings can be set during build and upstream and the build process does different things depending on what is set. The same frontend (ente-web) has to be hosted two times with different parameters for public albums to work (see commented link).

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:9]:
```suggestion
```

File: pkgs/by-name/en/ente-web/package.nix
Comment (doronbehar) [inline pkgs/by-name/en/ente-web/package.nix:13]:
```suggestion
```

**Review by doronbehar — COMMENTED** [review]
Looks good overall, but it won't be useful to anyone to merge this without a module added at the same PR.

File: pkgs/by-name/mu/museum/package.nix
Comment (nicegamer7) [inline pkgs/by-name/mu/museum/package.nix:17]: I think you should either remove the `v` prefix from here, or from the `version` attribute.

Comment (pinpox) [reply pkgs/by-name/mu/museum/package.nix:17]: fixed.

**Comment by pinpox** [comment]
> Looks good overall, but it won't be useful to anyone to merge this without a module added at the same PR.

Not sure I agree, you can just host it using a normal reverse-proxy. I will be submitting a module in a separate PR, but it is not ready yet.

File: pkgs/by-name/en/ente-web/package.nix
Comment (dotlambda) [inline pkgs/by-name/en/ente-web/package.nix:42]:
```suggestion
    runHook preInstall

    cp -r apps/photos/out $out

    runHook postInstall
```

**Review by dotlambda — COMMENTED** [review]
`meta.changelog` missing

**Comment by doronbehar** [comment]
> > Looks good overall, but it won't be useful to anyone to merge this without a module added at the same PR.
> 
> Not sure I agree, you can just host it using a normal reverse-proxy. I will be submitting a module in a separate PR, but it is not ready yet.

How is that going @pinpox ? I'm also very interested to know how exactly do you use this package, maybe we can decide on an alternative solution to this dilemma. 

I'd recommend you to decouple the `museum: fix version` commit, as we can and should merge it either way. Also, the rest of commits introducing the package `ente-web`, should be squashed to a single one. If a module will be added eventually, it should be added in a 2nd commit.

**Comment by doronbehar** [comment]
Please also note the nixfmt issues noted in the CI errors.

**Comment by pinpox** [comment]
@doronbehar I think it is a good idea to re-structure the package organisation a bit, since they all share the same source. I propose:
```nix
pkgs/by-name/en/ente/ente-web.nix
pkgs/by-name/en/ente/ente-server.nix
pkgs/by-name/en/ente/common.nix # for the src and shared stuff
```
I've seen this structure on a few other packages, do you have any objections?

**Comment by pinpox** [comment]
@doronbehar `ente-web`, `ente-server` and other ente stuff use the same `src` (apart from the sparse checkout difference). Would it be possible to unify the src or version so the packages are always in sync?

**Comment by doronbehar** [comment]
> @doronbehar `ente-web`, `ente-server` and other ente stuff use the same `src` (apart from the sparse checkout difference). Would it be possible to unify the src or version so the packages are always in sync?

Since they both use the same Git revision, then yes. But you'd have to not use the `sparseCheckout` argument, or perhaps put there both the values `"server"` and `"web"`. I am not experienced with that too, so you'd have to verify my anticipation.

**Comment by pinpox** [comment]
> I'd recommend you to decouple the `museum: fix version` commit,

Committed the fixed version as part of an update PR: https://github.com/NixOS/nixpkgs/pull/354212
I'll rebase this one as soon as it's merged

**Review by doronbehar — COMMENTED** [review]
OK this PR has been waiting for too long. We can merge this I guess without a module, but I don't think I'll be able to help reviewing the module. Should be good to go after these changes.

File: pkgs/by-name/en/ente-web/package.nix
Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:25]: I'm using `finalAttrs` instead of `rec` here, thinking this was the recommendation where. I'm not sure how to use `name` with that.

Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix:25]: > I'm using `finalAttrs` instead of `rec` here, thinking this was the recommendation where. I'm not sure how to use `name` with that.

Simple: `finalAttrs.src.name`

Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix:25]: @doronbehar When using the pattern with sourceRoot as above, it fails to build. I changed it like this:
```
  src = fetchFromGitHub {
    owner = "ente-io";
    repo = "ente";
    sparseCheckout = [ "web" ];
    rev = "photos-v${finalAttrs.version}";
    hash = "sha256-vucWtu7abQYFDKStAAzE2ofG87AZD2vEdNmPMnpQyeg=";
  };

  sourceRoot = "${finalAttrs.src.name}/web";
```
And get the following error:

```

syntax: prefetch-yarn-deps [path to yarn.lock] [options]

Options:
  -h --help         Show this help
  -v --verbose      Verbose output
  --builder         Only perform the download to current directory, then exit
error: builder for '/nix/store/sgahsqi02jirhcmny9xhsqwid2g0n0is-offline.drv' failed with exit code 1;
       last 12 log lines:
       > Running phase: patchPhase
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > no configure script, doing nothing
       > Running phase: buildPhase
       >
       > syntax: prefetch-yarn-deps [path to yarn.lock] [options]
       >
       > Options:
       >   -h --help         Show this help
       >   -v --verbose      Verbose output
       >   --builder         Only perform the download to current directory, then exit
       For full logs, run 'nix log /nix/store/sgahsqi02jirhcmny9xhsqwid2g0n0is-offline.drv'.
error: 1 dependencies of derivation '/nix/store/vc634i1h4cp0nrzbp1vyyphn20rw8jng-ente-web-0.9.16.drv' failed to build
```

Comment (doronbehar) [reply pkgs/by-name/en/ente-web/package.nix]: See the current implementation.

Comment (pinpox) [reply pkgs/by-name/en/ente-web/package.nix]: > See the current implementation.

Thanks for clarifying! I fixed the hash and rebuild the package, it builds correctly.

File: pkgs/by-name/mu/museum/package.nix
Comment (pinpox) [reply pkgs/by-name/mu/museum/package.nix]: Is this still open? Not sure what is missing.

Comment (doronbehar) [reply pkgs/by-name/mu/museum/package.nix]: > > Was this supposed to be updated in #354212 as well?
>
> Is this still open? Not sure what is missing.

If you don't touch the vendorHash of `museum/package.nix` in this PR, there is nothing to discuss about. But you do realize that in one of the revisions of _this_ PR, this change was present? This raises the suspicion that updating the vendor hash of the museum package is required after the update it had in #354212 - as usually both src and vendor hashes are updated in such version bump PRs.

Also, The fact that the `vendorHash` of `museum` is at the bottom of the expression, and not near the hash of the source, makes it easy to miss a lack of change to the `vendorHash`, which is almost always required in version bumps. Putting both hashes close to each other not only makes the diff slightly smaller and cleaner, but also guarantees that no reviewer will miss the fact that a vendor hash wasn't updated although the src _was_ updated.

**Review by doronbehar — COMMENTED** [review]
Could you squash the commits now? This is perhaps the 10th time I'm asking this.

File: pkgs/by-name/mu/museum/package.nix
Comment (pinpox) [reply pkgs/by-name/mu/museum/package.nix]: Let's keep this PR to ente-web then, I'l leave museum as is for now.

**Review by doronbehar — APPROVED** [review]

**Pull request merged**
